db/commitlog/commitlog.cc

```diff
@@ -431,7 +432,14 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     future<> rename_file(sstring, sstring) const;
     size_t max_request_controller_units() const;
     segment_id_type _ids = 0;
-    std::vector<sseg_ptr> _segments;
+
+    static constexpr size_t max_data_classes = size_t(data_class::max);
+
+    template<typename T>
+    using per_data_class = std::array<T, max_data_classes>;
+
+    per_data_class<std::vector<sseg_ptr>> _segments;    
+    per_data_class<std::optional<shared_future<with_clock<db::timeout_clock>>>> _segment_allocating;
```

Comment:
I see indeed one open segment per class. I see two problems with that:

1. even more write iops, reduces write batch size and increases write amplification. If lowfreq_lowvol is indeed very low frequency, then that shouldn't matter too much, but also bounds the gain.
2. the commitlog is no longer linear. We could restore linearity perhaps by recording an index and merging the two streams.

Reply:
How does it increase write ops? It increases memory buffer footprint, and potentially disk footprint as well, as we might have so low utilization in one queue that a segment might stay alive for a long time. But neither seems likely to really impact anything severely. Actual writes to disk should be the same (except maybe in the batch mode case, where we might increase # write ops per bytes written as we might fail to sneak in as much data in buffers. This needs to be investigated, but my spider sense tells me it too is negligible.

If by linear you mean RP:s, they only need to be monotonic per CF_ID, i.e. table. Which is why this is a designator per table. As long as we put data from a single table consequently in  either queue, we should be fine. Truncation etc is recorded per table. Btw, segments will be "linear" in creation still, they will just not be discarded and reused in order in-between queues.

Reply:
Let's say a shard sees 20 100-byte commits/task-quota. After the split we have 19 100-byte commits to one segment and 1 100-byte commit to another.

before: align_up(20*100, 1024) = 2048, 1 I/O
after: align_up(1*100, 1024) + align_up(19*100, 1024) = 3072, 2 I/O

1024 = xfs block size.

So we have 50% extra bytes and 100% extra I/O.

The numbers are made up but we did observe small commitlog writes (which prompted us to reduce the xfs block size and support sub-block writes).

Reply:
True, but in anything resembling "normal" write pressure I more or less assume we hit the end of buffers (even if we split them) more often than the period task removes them. I.e. aligning up and wasting would be fairly rare. Can't tell for sure, as this is really only run on a single benchmark.